### DISCUSSION TITLE
isync service (i.e. mbsync) stops functioning but produces no logs for debugging

### DISCUSSION THREAD

**Discussion by cacology** [Homebrew/discussions#4446] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: >=2.5.0 (shallow or no git repository)
ORIGIN: (none)
HEAD: (none)
Last commit: never
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: (none)
Core tap last commit: never
Core tap branch: (none)
Core tap JSON: 16 Apr 14:18 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.bk3593Jb5n/org.xquartz:0
HOMEBREW_EDITOR: /usr/local/opt/emacs-mac/bin/emacsclient -c
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: dodeca-core 64-bit arm_blizzard_avalanche
Clang: 14.0.3 build 1403
Git: 2.39.2 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3.1-arm64
CLT: N/A
Xcode: 14.3
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  pycparser
  cffi
```

### Description of issue

Calling `brew services start isync` as a non-admin user happily does what is expected: starts a instance of isync (i.e. mbsync) according to my files. However, after about 24 hours, it seems to stop working. I'd like to fix it so it doesn't periodically stall.

I've tried to debug it a bit, but cannot get it to produce log files. Can anyone advise?

## What I've tried

Restarting it when it stalls works fine, so I wanted to look at the log files. As it is, it sends out and error to `/dev/null` so I wrote my own plist, i.e.

homebrew.mxcl.isync-log.plist:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>EnvironmentVariables</key>
	<dict>
		<key>PATH</key>
		<string>/opt/homebrew/bin:/opt/homebrew/sbin:/usr/bin:/bin:/usr/sbin:/sbin</string>
	</dict>
	<key>Label</key>
	<string>homebrew.mxcl.isync</string>
	<key>LimitLoadToSessionType</key>
	<array>
		<string>Aqua</string>
		<string>Background</string>
		<string>LoginWindow</string>
		<string>StandardIO</string>
		<string>System</string>
	</array>
	<key>ProgramArguments</key>
	<array>
		<string>/opt/homebrew/opt/isync/bin/mbsync</string>
		<string>-a</string>
	</array>
	<key>RunAtLoad</key>
	<false/>
	<key>StandardErrorPath</key>
	<string>/Users/james/Library/Logs/homebrew.mxcl.isyncerr.log</string>
	<key>StandardOutPath</key>
	<string>/Users/james/Library/Logs/homebrew.mxcl.isyncout.log</string>
	<key>StartInterval</key>
	<integer>300</integer>
</dict>
</plist>
```

Then I loaded this with `brew services start isync --file=homebrew.mxcl.isync-log.plist`. It didn't seem to be running. I tried `/bin/launchctl bootstrap gui/503 /Users/james/Library/LaunchAgents/homebrew.mxcl.isync.plist` which complained: `Bootstrap failed: 5: Input/output error.`

How can I get a useful log file for this running as a non-admin user? Or, even better, any ideas why isync might be choking over night?

**Comment by pmiam** [comment]
I have a similar problem. Just started using MacOS this week. `brew service start isync` creates the LaunchAgent file as expected, but the service immediately fails with `brew services list` status `error 256`.

Attempting tobootstrap the service with launchctl fails with the same error `Bootstrap failed: 5: Input/output error` for both root and my user 501. however enabling/disabling does appear to change the `=>` status returned by `launchctl print-disabled gui/501`. This has no obvious effect.

**Reply by pmiam** [reply]
one possibility is that the daemon does not have access to my user's configuration file...

**Comment by tommasongr** [comment]
I experienced the same issue and I was able to output the logs by changing the plist like so:

```
<key>StandardErrorPath</key>
<string>/opt/homebrew/var/log/isync.log</string>
```

I then started the services with the command `brew services start isync --file=homebrew.mxcl.isync.plist`.

## Why isync was crashing

I also would like to share what was giving me the issue in the first place, so that someone else with a similar setup may have an easier time.

In my case, the error was caused by `pass` (which I use to store the imap password). Specifically I was calling it directly in the config file like so:

```shell
PassCmd "pass zoho.eu/neomutt | head -n 1"
```

Since I specified a custom location for the password store in my `.zprofile` (`export PASSWORD_STORE_DIR="$HOME/.passwords"`), the services wasn't able to find the password correctly.

The fix was pretty straightforward. I created a small shell script (`~/.config/isync-pass.sh`) to load the variable and run the command:

```shell
#!/usr/bin/env sh

eval "$(/opt/homebrew/bin/brew shellenv)"
export PASSWORD_STORE_DIR="$HOME/.passwords"

pass zoho.eu/neomutt | head -n 1
```

I then changed the isync config to use the script instead of the previous `pass` invocation:

```shell
PassCmd "~/.config/isync-pass.sh"
```

With this fix I was then able to start the service with the regular `brew services start isync` command. No custom plist required.
